cpp/ql/test/query-tests/Security/CWE/CWE-119/SAMATE/OverrunWriteProductFlow.expected

```diff
@@ -159,7 +147,5 @@ subpaths
 | test.cpp:203:9:203:15 | call to strncpy | test.cpp:147:19:147:24 | call to malloc | test.cpp:203:22:203:27 | string | This write may overflow $@ by 2 elements. | test.cpp:203:22:203:27 | string | string |
 | test.cpp:207:9:207:15 | call to strncpy | test.cpp:147:19:147:24 | call to malloc | test.cpp:207:22:207:27 | string | This write may overflow $@ by 3 elements. | test.cpp:207:22:207:27 | string | string |
 | test.cpp:243:5:243:10 | call to memset | test.cpp:241:20:241:38 | call to malloc | test.cpp:243:12:243:21 | string | This write may overflow $@ by 1 element. | test.cpp:243:16:243:21 | string | string |
-| test.cpp:243:5:243:10 | call to memset | test.cpp:241:27:241:32 | call to malloc | test.cpp:243:12:243:21 | string | This write may overflow $@ by 1 element. | test.cpp:243:16:243:21 | string | string |
```

Comment:
Perhaps you could explain what this PR does to me in a bit more detail by explaining why we got duplication here.  The corresponding code is:
```
    string_t str;
    char* buffer = (char*)malloc(size); // line 241
    set_string(&str, buffer);
    memset(str.string, 0, size + 1); // line 243
```

Reply:
Sure. Given an expression `e` we need to pick a unique dataflow node `n` for which `n.asExpr()` has a result (in order to prevent duplication). The node we pick is always the node containing the result of the fully converted expression. This is achieved on [this line](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll#L1077) which ensures that we don't give a result for `n.asExpr()` if `n` flows directly into a conversion (because in that case, it's not the fully converted expression).

The problem happens at this line:
```cpp
char* buffer = (char*)malloc(size); // line 241
```
The IR for this line looks like:
```
r0(glval<int *>)  = VariableAddress[buffer]      :
r1(glval<size_t>) = VariableAddress[size]        :
r2(size_t)        = Load[size]                   : &:r1, m4_6
r3(void *)        = Call[malloc]                 : func:r8_2, 0:r2
m1(unknown)       = ^InitializeDynamicAllocation : &:r3
r4(int *)         = Convert                      : r3
m2(int *)         = Store[buffer]                : &:r0, r4
```
we want the dataflow node `n` that wraps the instruction `r4` to give back `malloc(size)` when we call `n.asExpr()`. However, the `unique` I linked to above doesn't do that in this case since the `r3` flows to _two_ destinations:
- To `m1(unknown)       = ^InitializeDynamicAllocation : &:r3`
- To `r4(int *)         = Convert                      : r3`

The `InitializeDynamicAllocation` instruction is a side effect instruction that records that a value was written to what `malloc` points to. However, since they both consume `r3` we don't flow uniquely to `r4`, and thus `n.asExpr()` won't have a result for `r4`. Instead, it would have a result for `r3` instead (and for technical reasons _also_ from `r4` since that's the instruction the IR correctly says contains the result of `(char*)malloc(size)`).

So by excluding `InitializeDynamicAllocation` from that `unique` we ensure that we flow to `r4` so that we identify that the dataflow node `n` that wraps `r4` should be the node for which `n.asExpr()` returns `malloc(size)`.

Does that make sense?

Reply:
I think so, yes, thanks!